### ISSUE TITLE
Fix Typescript type inference for `returning()` to better support wildcard ("*") calls

### ISSUE THREAD

**Issue by CodyEakins** [tgriesser/knex#3444]
# Environment

Knex version: 0.19.3
Database + version: Postgres 11
OS: Windows 10, WSL

@lorefnon - Hopefully you can address this... Or at least tell me if I am wrong.

# Bug

When using `.returning()` in a Query Builder chain, it's possible to pass in a wildcard ("*") to indicate that you want all of a table's columns returned with the result.

For example, when using Typescript Generics:
```ts
interface User {
  id: number;
  email: string;
  password: string;
}

const queryResult = await knex<User>("user").insert({
  email: "name@example.com",
  password: "passwordHash"
}).returning("*");
```

One expects `queryResult`'s type to be `User[]` here, but it is actually `Partial<User>[]`. 

This result makes sense when calling `returning()` with a column identifier (e.g. - `returning("id")`), but not when it is called with a wildcard. When called with a wildcard, the result set is an array of fully qualified `User` objects, not partial ones.

---

**Note:**

I know you can pass the `User` type to `returning()` more directly, like so:
```ts
.returning<User>("*")
```
And get the expected result... But this bug is more about the type inference from the main Query Builder to `returning()`.

---

**Suggested Fix:**

Maybe update the definition to something like: 
```ts
returning<TResult2 = SafePartial<TRecord>[] | TRecord[]>(
  column: string | string[]
): QueryBuilder<TRecord, TResult2>;
```

**Comment by KristjanTammekivi** [comment]
Mildly OT but maybe enable a zero-parameter .returning() which automatically means return all columns

**Comment by lorefnon** [comment]
https://github.com/tgriesser/knex/pull/3446 adds overrides that special case '*' so that the returned value will be the complete record. 

> Maybe update the definition to something like:

> returning<TResult2 = SafePartial<TRecord>[] | TRecord[]>(
  column: string | string[]
): QueryBuilder<TRecord, TResult2>;

This is unlikely to get you what you want because `T extends Partial<T>` and hence `T | Partial<T>` is effectively `Partial<T>`.

**Comment by lorefnon** [comment]
I also noticed that travis no longer run `lint:types`. Was this an intentional change ? cc. @kibertoad

**Comment by kibertoad** [comment]
@lorefnon Huh, I wonder how did that happen. Opened #3450 to address that.

**Comment by CodyEakins** [comment]
Thanks for the fast and thorough response to this.

**Comment by kibertoad** [comment]
Released in 0.19.5
